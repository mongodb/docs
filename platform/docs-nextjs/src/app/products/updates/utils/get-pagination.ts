/**
 *
 * @param array - flatten array
 * @param chunkSize - max size of each array
 * @returns [[{}, {}], [{}]]
 * @description Takes an flatten array and creates one level nested array of objects based on the chunkSize. Used for pagination data.
 */
export const getPagination = <T>(array: T[], chunkSize: number): T[][] => {
  const chunks: T[][] = [];
  if (!array.length) {
    return [[]]; // default to one empty nested array
  }
  for (let i = 0; i < array.length; i += chunkSize) {
    chunks.push(array.slice(i, i + chunkSize));
  }
  return chunks;
};
