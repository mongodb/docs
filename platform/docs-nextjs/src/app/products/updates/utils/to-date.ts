export const normalizeDate = (date: string) => {
  // Check if the date string has the correct format 'YYYY-MM-DD'
  const dateToNormalize = /^(\d{4})-(\d{2})-(\d{2})$/.exec(date)?.slice(1); // excludes the full date in the first index i.e. "2025-05-27"
  if (dateToNormalize) {
    // This is needed because, JavaScript’s Date constructor uses zero-based months (January = 0 … December = 11).
    return new Date(Number(dateToNormalize[0]), Number(dateToNormalize[1]) - 1, Number(dateToNormalize[2]));
  }
  return new Date(date);
};
