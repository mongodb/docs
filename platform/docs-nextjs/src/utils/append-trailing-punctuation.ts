import { ASTNode, ReferenceNode, RefRoleNode } from "@/types/ast";
import { isReferenceNode, isRefRoleNode, isTextNode } from "@/types/ast-utils";

// Append dangling punctuation to preceding Link's text
export const appendTrailingPunctuation = (nodes: ASTNode[]) => {
  const truncatedNodes = [];

  for (let i = 0; i < nodes.length; i++) {
    const currNode = nodes[i];
    const nextNode = nodes[i + 1];
    const hasDanglingSibling =
      isTextNode(nextNode) &&
      nextNode.value.length === 1 &&
      !nextNode.value.match(/\s/);
    // TODO: if issues arise, instead of matching for non-whitespace, we might consider checking FOR certain singular punctuation
    // ie: regex match for  .?!,:;)*"']}`

    if ((isReferenceNode(currNode) || isRefRoleNode(currNode)) && hasDanglingSibling) {
      const copyOfNodeWithDeepCopiedChildren = deepCopyNodesChildren(currNode);
      copyOfNodeWithDeepCopiedChildren.children.push(nextNode);
      truncatedNodes.push(copyOfNodeWithDeepCopiedChildren);
      i++;
      continue;
    }
    truncatedNodes.push(currNode);
  }
  return truncatedNodes;
};

// Make a copy of node with a deep copy of the new child node
function deepCopyNodesChildren(refNode: ReferenceNode | RefRoleNode) {
  if (!refNode.children || !refNode.children.length) return { ...refNode };
  const copyOfChildren = [...refNode.children];

  const copyOfNode = {
    ...refNode,
    children: copyOfChildren,
  };
  return copyOfNode;
}
