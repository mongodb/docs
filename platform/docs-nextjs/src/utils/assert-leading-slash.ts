const assertLeadingSlash = (url: string) => {
  if (url && url.match(/^\//)) {
    return url;
  }
  return `/${url}`;
};

export { assertLeadingSlash };
