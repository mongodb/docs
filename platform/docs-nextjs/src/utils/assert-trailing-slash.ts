export const assertTrailingSlash = (url: string) => {
  if (url && url.match(/\/$/)) {
    return url;
  }
  return `${url}/`;
};
