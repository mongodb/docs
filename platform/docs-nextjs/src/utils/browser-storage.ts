import { isBrowser } from './is-browser';

const isValidStorage = isBrowser;

export const setLocalValue = (key: string, value: unknown) => {
  try {
    if (isValidStorage) {
      const prevState = JSON.parse(window.localStorage.getItem('mongodb-docs') || '{}');
      localStorage.setItem('mongodb-docs', JSON.stringify({ ...prevState, [key]: value }));
    }
  } catch {
    console.error('Error setting localStorage value');
  }
};

export const getLocalValue = (key: string) => {
  try {
    if (isValidStorage && JSON.parse(window.localStorage.getItem('mongodb-docs') || '{}')) {
      const docsObj = JSON.parse(window.localStorage.getItem('mongodb-docs') || '{}');
      if (docsObj) {
        return docsObj[key];
      }
    }
    return undefined;
  } catch {
    console.error('Error getting localStorage value');
    return undefined;
  }
};

export const setSessionValue = (key: string, value: string) => {
  try {
    if (isValidStorage) {
      const prevState = JSON.parse(window.sessionStorage.getItem('mongodb-docs') || '{}');
      sessionStorage.setItem('mongodb-docs', JSON.stringify({ ...prevState, [key]: value }));
    }
  } catch {
    console.error('Error setting session value');
  }
};

export const getSessionValue = (key: string) => {
  try {
    if (isValidStorage && JSON.parse(window.sessionStorage.getItem('mongodb-docs') || '{}')) {
      const docsObj = JSON.parse(window.sessionStorage.getItem('mongodb-docs') || '{}');
      if (docsObj) {
        return docsObj[key];
      }
    }
    return undefined;
  } catch {
    console.error('Error getting session value');
    return undefined;
  }
};
