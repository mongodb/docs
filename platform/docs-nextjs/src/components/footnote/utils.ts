import type { ASTNode } from '@/types/ast';
import type { FootnoteNode, FootnoteReferenceNode } from '@/types/ast';
import { findAllKeyValuePairs } from '@/utils/find-all-key-value-pairs';
import type { FootnoteType } from './footnote-context';

export const getFootnotes = (nodes: ASTNode[]) => {
  const footnotes = findAllKeyValuePairs(nodes, 'type', 'footnote') as FootnoteNode[];
  const footnoteReferences = findAllKeyValuePairs(nodes, 'type', 'footnote_reference') as FootnoteReferenceNode[];
  const numAnonRefs = footnoteReferences.filter(
    (node) => !Object.prototype.hasOwnProperty.call(node, 'refname'),
  ).length;
  // We label our footnotes by their index, regardless of their names to
  // circumvent cases such as [[1], [#], [2], ...]
  return footnotes.reduce<Record<string, FootnoteType>>((map, footnote, index) => {
    if (footnote.name) {
      // Find references associated with a named footnote

      map[footnote.name] = {
        label: index + 1,
        references: getNamedFootnoteReferences(footnoteReferences, footnote.name),
      };
    } else {
      // Find references associated with an anonymous footnote
      // Replace potentially broken anonymous footnote ids
      footnote.id = `${index + 1}`;

      map[footnote.id] = {
        label: index + 1,
        references: getAnonymousFootnoteReferences(index, numAnonRefs),
      };
    }
    return map;
  }, {});
};

// Find all footnote_reference node IDs associated with a given footnote by
// that footnote's refname
const getNamedFootnoteReferences = (footnoteReferences: FootnoteReferenceNode[], refname: string) => {
  return footnoteReferences.filter((node) => node.refname === refname).map((node) => node.id);
};

// They are used infrequently, but here we match an anonymous footnote to its reference.
// The nth footnote on a page is associated with the nth reference on the page. Since
// anon footnotes and footnote references are anonymous, we assume a 1:1 pairing, and
// have no need to query nodes. If there are more anonymous footnotes than references,
// we may return an empty array
const getAnonymousFootnoteReferences = (index: number, numAnonRefs: number) => {
  return index > numAnonRefs ? [] : [`id${index + 1}`];
};
