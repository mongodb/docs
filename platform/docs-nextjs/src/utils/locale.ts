import { SIDE_NAV_CONTAINER_ID, TEMPLATE_CONTAINER_ID } from '@/constants';
import { assertTrailingSlash } from './assert-trailing-slash';
import { isBrowser } from './is-browser';
import { normalizePath } from '@/utils/normalize-path';
import { removeLeadingSlash } from '@/utils/remove-leading-slash';

export type AvailableLocaleType = 'en-us' | 'pt-br' | 'es' | 'ko-kr' | 'ja-jp' | 'it-it' | 'de-de' | 'fr-fr' | 'zh-cn';

export type AvailableLanguageData = {
  language: string;
  localeCode: AvailableLocaleType;
  fontFamily?: string;
};

// Prevents Smartling from translating content within HTML block
// (https://help.smartling.com/hc/en-us/articles/13274689281307-Default-Translatable-Content-In-The-GDN)
export const NOTRANSLATE_CLASS = 'notranslate';

/**
 * Key used to access browser storage for user's preferred locale
 */
export const STORAGE_KEY_PREF_LOCALE = 'preferredLocale';
// Should be the same as what B2K expects
export const COOKIE_KEY_PREF_LOCALE = 'mdb_docsPrefLocale';

// Update this as more languages are introduced
// Because the client-side redirect script cannot use an import, PLEASE remember to update the list of supported languages
// in redirect-based-on-lang.js
const AVAILABLE_LANGUAGES: AvailableLanguageData[] = [
  { language: 'English', localeCode: 'en-us' },
  { language: '简体中文', localeCode: 'zh-cn', fontFamily: 'Noto Sans SC' },
  { language: '한국어', localeCode: 'ko-kr', fontFamily: 'Noto Sans KR' },
  { language: 'Português', localeCode: 'pt-br' },
  { language: '日本語', localeCode: 'ja-jp', fontFamily: 'Noto Sans JP' },
];

// GTM locale
export const BETA_LOCALE: { [key: string]: AvailableLanguageData } = {
  es: {
    language: 'Español',
    localeCode: 'es',
  },
};

// Languages in current development that we do not want displayed publicly yet
const HIDDEN_LANGUAGES: AvailableLanguageData[] = [BETA_LOCALE['es']];

/**
 * @param {boolean} forceAll - Bypasses feature flag requirements if necessary
 * @returns An array of languages supported for translation
 */
export const getAvailableLanguages = (forceAll = false) => {
  const langs = [...AVAILABLE_LANGUAGES];

  if (forceAll || process.env.GATSBY_FEATURE_SHOW_HIDDEN_LOCALES === 'true') {
    langs.push(...HIDDEN_LANGUAGES);
  }

  return langs;
};

const validateLocaleCode = (potentialCode: string) =>
  // Include hidden languages in validation to ensure current locale of hidden sites can still be captured correctly
  !!getAvailableLanguages(true).find(({ localeCode }) => potentialCode === localeCode);

/**
 * Strips the first locale code found in the slug. This function should be used to determine the original un-localized path of a page.
 * This assumes that the locale code is the first part of the URL slug. For example: "/zh-cn/docs/foo".
 * @param {string} slug
 * @returns {string}
 */
const stripLocale = (slug: string) => {
  // Smartling has extensive replace logic for URLs and slugs that follow the pattern of "https://www.mongodb.com/docs". However,
  // there are instances where we can't rely on them for certain components
  if (!slug) {
    return '';
  }

  // Normalize the slug in case any malformed slugs appear like: "//zh-cn/docs"
  const slugWithSlash = slug.startsWith('/') ? slug : `/${slug}`;
  const normalizedSlug = normalizePath(slugWithSlash);
  const firstPathSlug = normalizedSlug.split('/', 2)[1];

  // Replace from the original slug to maintain original form
  const res = validateLocaleCode(firstPathSlug) ? normalizePath(slug.replace(firstPathSlug, '')) : slug;
  if (res.startsWith('/') && !slug.startsWith('/')) {
    return removeLeadingSlash(res);
  } else if (!res.startsWith('/') && slug.startsWith('/')) {
    return `/${res}`;
  } else {
    return res;
  }
};

export const getAllLocaleCssStrings = () => {
  const strings: string[] = [];
  // We want to bypass feature flag requirements to ensure fonts for hidden languages are always included
  const allLangs = getAvailableLanguages(true);

  allLangs.forEach(({ localeCode, fontFamily }) => {
    if (!fontFamily) {
      return;
    }
    const [languageCode] = localeCode.split('-');
    // Only check that languageCode is in the beginning to be flexible when region code is capitalized
    // For example: zh-cn and zh-CN will be treated the same.
    // We want to target everything except for inline code, code blocks, and the consistent-nav components
    strings.push(`
      html[lang^=${languageCode}] {
        #${TEMPLATE_CONTAINER_ID} *:not(:is(code, code *)),
        #${SIDE_NAV_CONTAINER_ID} * {
          font-family: ${fontFamily};
        }

        // Italicized non-latin characters may look confusing, so we want to replace them with bold
        // without changing the source HTML tag (and potentially causing errors with Smartling)
        em,
        h1 .guilabel,
        h2 .guilabel,
        h3 .guilabel,
        h4 .guilabel,
        h5 .guilabel,
        h6 .guilabel {
          font-style: normal;
          font-weight: bold;
        }
      }
    `);
  });

  return strings;
};

/**
 * Returns the locale code based on the current location pathname of the page.
 */
export const getCurrLocale = (): AvailableLocaleType => {
  const defaultLang = 'en-us';

  if (!isBrowser) {
    return defaultLang;
  }

  // This currently needs to be client-side because the source page doesn't know about locale at
  // build time. Smartling's GDN handles localization
  // Example on https://www.mongodb.com/zh-cn/docs/manual/introduction:
  // expected pathname - /zh-cn/docs/manual/introduction; expected locale - "zh-cn"
  const pathname = window.location.pathname;
  const expectedDocsPrefixes = ['docs', 'docs-qa'];
  const firstPathSlug = pathname.split('/', 2)[1];
  if (expectedDocsPrefixes.includes(firstPathSlug)) {
    return defaultLang;
  }

  const slugMatchesCode = validateLocaleCode(firstPathSlug);
  return slugMatchesCode ? (firstPathSlug as AvailableLocaleType) : defaultLang;
};

/**
 * Returns the pathname with its locale code prepended. Leading slashes are preserved, if they exist.
 * @param {string} pathname - Path name or slug of the page
 * @param {string?} localeCode - Optional locale code. By default, the code is determined based on the current location of the page
 */
export const localizePath = (pathname: string, localeCode?: string) => {
  if (!pathname) {
    return '';
  }

  const unlocalizedPath = stripLocale(pathname);
  const code = localeCode && validateLocaleCode(localeCode) ? localeCode : getCurrLocale();
  const languagePrefix = code === 'en-us' ? '' : `${code}/`;
  let newPath = languagePrefix + unlocalizedPath;
  if (pathname.startsWith('/')) {
    newPath = `/${newPath}`;
  }
  return normalizePath(newPath);
};

/**
 * Returns a mapping of a page's URL and its equivalent URLs for different languages.
 * @param {string} siteUrl
 * @param {string} fullPath
 * @returns {object}
 */
export const getLocaleMapping = (siteUrl: string, fullPath: string) => {
  // handle the `/` path
  const slugForUrl = fullPath === '/' ? '' : fullPath;
  const normalizedSiteUrl = siteUrl?.endsWith('/') ? siteUrl.slice(0, -1) : siteUrl;
  const localeHrefMap: { [k: string]: string } = {};

  getAvailableLanguages().forEach(({ localeCode }) => {
    const localizedPath = localizePath(slugForUrl, localeCode);
    const targetUrl = normalizedSiteUrl + localizedPath;
    localeHrefMap[localeCode] = assertTrailingSlash(targetUrl);
  });

  return localeHrefMap;
};

export const onSelectLocale = (locale: string) => {
  try {
    // Set cookie to expire in 1 year
    const maxAge = 60 * 60 * 24 * 365;
    const cookieMaxAge = `max-age=${maxAge}`;
    // Set for all paths on the MongoDB domain so that it persists for every site and locale
    const cookiePath = 'path=/';
    window.document.cookie = `${COOKIE_KEY_PREF_LOCALE}=${locale};${cookieMaxAge};${cookiePath}`;
  } catch (err) {
    console.error(err);
  }

  const location = window.location;
  const localizedPath = localizePath(location.pathname, locale);
  window.location.pathname = localizedPath;
};

/**
 * Ensures that a locale code has an all lowercase language code with an all uppercase region code,
 * as described in https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/lang#region_subtag.
 * @param {string} localeCode - A valid locale code with either 1 or 2 parts. Example: `zh-cn` or `zh`
 */
export const getHtmlLangFormat = (localeCode: string) => {
  const parts = localeCode.split('-');
  if (parts.length < 2) {
    return localeCode;
  }

  const [langCode, regionCode] = parts;
  return `${langCode.toLowerCase()}-${regionCode.toUpperCase()}`;
};
