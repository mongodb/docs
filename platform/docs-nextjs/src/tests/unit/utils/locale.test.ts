import { mockLocation } from '@/tests/utils/mock-location';
import { getAvailableLanguages, getHtmlLangFormat, getLocaleMapping, localizePath } from '@/utils/locale';

describe('getLocaleMapping', () => {
  it.each([
    ['https://www.mongodb.com/docs/', '/'],
    ['https://www.mongodb.com/docs', '/about/page/'],
    ['https://www.mongodb.com', 'introduction'],
  ])('returns a valid mapping of URLs', (siteUrl, slug) => {
    const mapping = getLocaleMapping(siteUrl, slug);
    expect(Object.keys(mapping)).toHaveLength(getAvailableLanguages().length);
    expect(mapping).toMatchSnapshot();
  });
});

describe('localizePath', () => {
  it.each([
    ['/', '/zh-cn/'],
    ['/page/slug', '/zh-cn/page/slug'],
    ['page/slug/', 'zh-cn/page/slug/'],
  ])('returns localized path when no code is set', (slug, expectedRes) => {
    // Pretend page exists on translated site
    mockLocation({ pathname: '/zh-cn/docs-test/drivers/node/current' });
    const res = localizePath(slug);
    expect(res).toEqual(expectedRes);
  });

  it.each([
    ['/', 'ko-kr', '/ko-kr/'],
    ['/page/slug', 'pt-br', '/pt-br/page/slug'],
    ['page/slug/', 'zh-cn', 'zh-cn/page/slug/'],
  ])('returns localized path when code is set', (slug, code, expectedRes) => {
    const res = localizePath(slug, code);
    expect(res).toEqual(expectedRes);
  });

  it.each([
    ['/', '/'],
    ['/page/slug', '/page/slug'],
    ['page/slug/', 'page/slug/'],
  ])('returns the same page slug when English is found by default', (slug, expectedRes) => {
    // Simulate English docs site
    mockLocation({ pathname: '/docs/manual' });
    const res = localizePath(slug);
    expect(res).toEqual(expectedRes);
  });

  it.each([
    ['/', '/'],
    ['/page/slug', '/page/slug'],
    ['page/slug/', 'page/slug/'],
  ])('gracefully defaults to English path when invalid language is passed in', (slug, expectedRes) => {
    // Simulate English docs site
    mockLocation({ pathname: '/docs/manual' });
    const res = localizePath(slug, 'beep-boop');
    expect(res).toEqual(expectedRes);
  });

  it.each([
    ['zh-cn/docs', 'zh-cn/docs'],
    ['/ko-kr/docs/foo', '/zh-cn/docs/foo'],
    ['/pt-br/', '/zh-cn/'],
    ['//ko-kr/', '/zh-cn/'],
    ['pt-br/', 'zh-cn/'],
    ['pt-br/docs/pt-br', 'zh-cn/docs/pt-br'],
  ])('handles slugs that already have a locale code', (slug, expectedRes) => {
    const res = localizePath(slug, 'zh-cn');
    expect(res).toEqual(expectedRes);
  });
});

describe('getHtmlLangFormat', () => {
  it.each([
    ['zh-cn', 'zh-CN'],
    ['zh-CN', 'zh-CN'],
    ['zh', 'zh'],
  ])('transforms %s into %s', (input, expectedRes) => {
    const res = getHtmlLangFormat(input);
    expect(res).toEqual(expectedRes);
  });
});
