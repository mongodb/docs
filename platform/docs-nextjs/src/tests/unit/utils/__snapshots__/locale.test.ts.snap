// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`getLocaleMapping returns a valid mapping of URLs 1`] = `
{
  "en-us": "https://www.mongodb.com/docs/",
  "ja-jp": "https://www.mongodb.com/docs/",
  "ko-kr": "https://www.mongodb.com/docs/",
  "pt-br": "https://www.mongodb.com/docs/",
  "zh-cn": "https://www.mongodb.com/docs/",
}
`;

exports[`getLocaleMapping returns a valid mapping of URLs 2`] = `
{
  "en-us": "https://www.mongodb.com/docs/about/page/",
  "ja-jp": "https://www.mongodb.com/docs/ja-jp/about/page/",
  "ko-kr": "https://www.mongodb.com/docs/ko-kr/about/page/",
  "pt-br": "https://www.mongodb.com/docs/pt-br/about/page/",
  "zh-cn": "https://www.mongodb.com/docs/zh-cn/about/page/",
}
`;

exports[`getLocaleMapping returns a valid mapping of URLs 3`] = `
{
  "en-us": "https://www.mongodb.comintroduction/",
  "ja-jp": "https://www.mongodb.comja-jp/introduction/",
  "ko-kr": "https://www.mongodb.comko-kr/introduction/",
  "pt-br": "https://www.mongodb.compt-br/introduction/",
  "zh-cn": "https://www.mongodb.comzh-cn/introduction/",
}
`;
